app/src/main/java/org/mozilla/focus/utils/NewFeatureNotice.java

```diff
@@ -0,0 +1,87 @@
+package org.mozilla.focus.utils;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.support.annotation.VisibleForTesting;
+
+public class NewFeatureNotice {
+
+    private static final String PREF_KEY_BOOLEAN_FIRSTRUN_SHOWN = "firstrun_shown";
+    private static final String PREF_KEY_INT_FEATURE_UPGRADE_VERSION = "firstrun_upgrade_version";
+    private static final int MULTI_TAB_FROM_VERSION_1_0_TO_2_0 = 1;
+    private static final int FIREBASE_FROM_VERSION_2_0_TO_2_1 = 2;
+
+    private static NewFeatureNotice instance;
+
+    private final SharedPreferences preferences;
+
+    public synchronized static NewFeatureNotice getInstance(Context context) {
+        if (instance == null) {
+            instance = new NewFeatureNotice(context.getApplicationContext());
+        }
+        return instance;
+    }
+
+    private NewFeatureNotice(Context context) {
+        preferences = PreferenceManager.getDefaultSharedPreferences(context);
+    }
+
+    public boolean shouldShowMultiTabUpdate() {
+        return MULTI_TAB_FROM_VERSION_1_0_TO_2_0 == getLastShownFeatureVersion() + 1;
+    }
+
+    public void setMultiTabUpdateNoticeDidShow() {
+        setFirstRunDidShow();
+        setLastShownFeatureVersion(MULTI_TAB_FROM_VERSION_1_0_TO_2_0);
+    }
+
+    public boolean shouldShowPrivacyPolicyUpdate() {
+        if (isNewlyInstalled()) {
+            setPrivacyPolicyUpdateNoticeDidShow();
+            return false;
+        }
+
+        return FIREBASE_FROM_VERSION_2_0_TO_2_1 == getLastShownFeatureVersion() + 1;
+    }
+
+    public void setPrivacyPolicyUpdateNoticeDidShow() {
+        setLastShownFeatureVersion(FIREBASE_FROM_VERSION_2_0_TO_2_1);
```

Comment:
I think this may be override to v1 if the user first tap the privacy notice but later come back to finish 1.0 -> 2.0 onboarding?

Reply:
I think adding an extra privacy notice shared preference would be simpler.

Reply:
1. That won't happen. The notification would only show after the 1.0 => 2.0 onboarding is done protected by the statement below in `shouldShowPrivacyPolicyUpdate()`
`return FIREBASE_FROM_VERSION_2_0_TO_2_1 == getLastShownFeatureVersion() + 1;`

2. Yes, it would be more simple to code that way, however, we need to keep adding new preference for the new feature hint each time in the future. The preference could finally blow up.  So. I try reusing the same preference for the app upgrade kind task.

Reply:
I just tried again and I noticed after onboarding is shown we'll receive the notice. I see why you're coding these this way now.

Reply:
I'm thinking of privacy notice as not a version of `NewFeatureNotice` since it's a Notification rather then a new Onboarding content, so I think it is this nature of different displays that make me feel like separating them could be simpler. (Notification v1, Onboarding v2...etc) But current implementation is better in a way that we only display one thing at a time by forcing user to upgrade once at a time. (No simultaneous Notification and Onboarding) I'm just a bit worried about since getting more and more complex when we're having more upgrade versions and some of them should always be shown, while some of them should be skipped. These protection need to be handled very carefully so the old (v1, v2 when v3 is coming) messages will work correctly.